share/picongpu/tests/CurrentDeposition/lib/python/test/CurrentDeposition/MainTest.py

```diff
@@ -0,0 +1,232 @@
+        pos_offs(list): all node positions of the particle in the series
+        charge(float): charge of particle
+
+    """
+    num_iter = len(series.iterations)
+
+    # create arrays for saving the data
+    Js = []
+    poss = []
+    pos_offs = []
+
+    for iteration in range(0, num_iter, 1):
+        i = series.iterations[iteration]  # current iteration
+
+        # current density field
+        J_x = i.meshes["J"]["x"].load_chunk()
+        J_y = i.meshes["J"]["y"].load_chunk()
+        J_z = i.meshes["J"]["z"].load_chunk()
+
+        # charge
+        charge = i.particles["e"]["charge"][io.Mesh_Record_Component.SCALAR]
+
+        # node coordinates
+        pos2_x = i.particles["e"]["position"]["x"].load_chunk()
+        pos2_y = i.particles["e"]["position"]["y"].load_chunk()
+        pos2_z = i.particles["e"]["position"]["z"].load_chunk()
+
+        # InCell coordinates
+        pos_off2_x = i.particles["e"]["positionOffset"]["x"].load_chunk()
+        pos_off2_y = i.particles["e"]["positionOffset"]["y"].load_chunk()
+        pos_off2_z = i.particles["e"]["positionOffset"]["z"].load_chunk()
+
+        series.flush()
+
+        # write coordinate tuple
+        pos2 = np.array([*pos2_x, *pos2_y, *pos2_z])
+        pos_off = np.array([*pos_off2_x, *pos_off2_y, *pos_off2_z])
+
+        # write current density tuple
+        J = [J_x, J_y, J_z]
+        Js.append(J)
+        poss.append(pos2)
+        pos_offs.append(pos_off)
+
+    return Js, poss, pos_offs, charge
+
+
+def compare(j_grid_x, j_grid_y, j_grid_z, J):
+    """Compares the current density J provided by the simulation with the
+    values j_grid_x/y/z from the python reference implementation.
+    herefore abs(j(PIConGPU) - j(Python)) is calculated and compared against
+    epsilon = 1e-5, which is slightly higher than PIConGPU numerical
+    uncertainty.
+
+    Parameters:
+    j_grid_x/y/z(np.array): current density calculated by the reference
+    J(np.array): current density calculated by PIConGPU
+
+    Returns:
+    0: if both current densities coincide
+    42: if they don't coincide
+    43: if the masked arrays _not_zero don't have the same shape
+
+    """
+
+    # check if the shapes of algo_not zero is the same as sim_not_zero
+    try:
+        # compare the arrays values by dividing; -1 for np.where operation
+        x_compare = abs(J[0] - j_grid_x)
+        y_compare = abs(J[1] - j_grid_y)
+        z_compare = abs(J[2] - j_grid_z)
+
+        # error boundary; slightly higher than numerical uncertainty
+        epsilon = 1e-5
+
+        if (np.all(x_compare < epsilon) and np.all(y_compare < epsilon) and
+                np.all(z_compare < epsilon)):
+            print("simulation and reference coincide")
+            return 0
+        else:
+            print("no consensus between simulation and reference")
+            print("please check the simulation")
+            return 42
+
+    except ValueError:
```

Comment:
Make the exception catching as narrow as possible. I presume that this is supposed to catch potential broadcasting issues in `x_compare = ...` etc.? If so, move the `except` clause directly up there. If you also want to guard against something broader, also make the printed message broader and consider using multiple `except` clauses.

Reply:
Thanks for the tip. This statement comes from an older version of the test. There I compared only the non-zero entries J[J != 0] by decision with the non-zero reference values. It then was possible that the two non-zero arrays had not the same shape and therefore the dividing operation would have yielded a ValueError I wanted to catch.
This is not the case anymore, since the reference current density array has now the same shape as the simulation's one. So it is not necessary to implement this exception and I would remove it.

Reply:
Great, thanks! (I would love to just `resolve conversation` but I can't so you'll get a nice comment from me for every resolved conversation instead. =) )